File: 257. Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> path;
        if(!root) return path;
        DFS(path, root, to_string(root->val));
        return path;
    }

    void DFS(vector<string>& path, TreeNode* root, string curr){
        if(!root) return;
        if(!root->left && !root->right){
            path.push_back(curr);
            return;
        }
        if(root->left) DFS(path, root->left, curr + "->" + to_string(root->left->val));
        if(root->right) DFS(path, root->right, curr + "->" + to_string(root->right->val));
    }

    // dfs: O(N), 17.33%, ok
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ret;
        if(!root) return ret;
        dfs(root, "", ret);
        return ret;
    }

    void dfs(TreeNode* root, string path, vector<string>& ret){
        if(!root->left && !root->right){
            path += "->" + to_string(root->val);
            ret.push_back(path.substr(2));
        }
        if(root->left) dfs(root->left, path + "->" + to_string(root->val), ret);
        if(root->right) dfs(root->right, path + "->" + to_string(root->val), ret);
    }
};

File: 257. Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # recursive dfs: O(N) time, O(h) space
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        def dfs(root, curr):
            if not root.left and not root.right:
                self.ret.append(curr + str(root.val))
                return 
            if root.left:
                dfs(root.left, curr + str(root.val) + '->')
            if root.right:
                dfs(root.right, curr + str(root.val) + '->')
        self.ret = []
        if root is None:
            return self.ret
        dfs(root, "")
        return self.ret
    
    # bfs queue: O(N) time, O(N) space
    def binaryTreePaths(self, root):
        ret, queue = [], [(root, '')]
        if not root:
            return ret
        for elem in queue:
            node = elem[0]
            if not node.left and not node.right:
                ret.append(elem[1] + str(node.val))
            queue += [(n, elem[1] + str(node.val) + '->') for n in [node.left, node.right] if n]
        return ret
        